## DATA EXPLORATION

This section shows the inital data exploration and cleaning.

In [1]:
import pandas as pd

In [2]:
data = pd.read_csv('../data/raw/historical_stock.csv', index_col=0, header=[0,1])

data.columns = [f"{col[0]}_{col[1]}" for col in data.columns]

df = pd.DataFrame(data)

df.head(10)

,SIE.DE_Close,SAN_Close,AIR.PA_Close,ITX.MC_Close,IBE.MC_Close
Date,,,,,
2022-01-03,135.098312,2.880033,106.550911,24.326015,8.459363
2022-01-04,136.592377,2.940033,109.064682,24.198698,8.410771
2022-01-05,138.824677,2.914319,110.459190,24.156258,8.354075
2022-01-06,135.485031,3.034320,108.844498,23.994993,8.301434
2022-01-07,133.727295,3.068606,107.982109,23.477236,8.305483
2022-01-10,129.965790,3.077177,108.183937,23.494213,8.251952
2022-01-11,132.092621,3.094321,108.642654,23.714895,8.256070
2022-01-12,133.780045,3.085749,108.881203,23.672455,8.227246
2022-01-13,133.147247,3.085749,107.615143,23.663963,8.243717


In [3]:
df.info() 
df.isnull().sum()

<class 'pandas.DataFrame'>
Index: 775 entries, 2022-01-03 to 2024-12-31
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   SIE.DE_Close  766 non-null    float64
 1   SAN_Close     753 non-null    float64
 2   AIR.PA_Close  768 non-null    float64
 3   ITX.MC_Close  768 non-null    float64
 4   IBE.MC_Close  768 non-null    float64
dtypes: float64(5)
memory usage: 36.3+ KB


SIE.DE_Close     9
SAN_Close       22
AIR.PA_Close     7
ITX.MC_Close     7
IBE.MC_Close     7
dtype: int64

The *historical_stock.csv* file has 775 entries, from **2022-01-01** to **2025-01-01**. It is easyly observable that at least 7 days are null in every column, those days need to be assessed, as well as some other faulty registries clearly shown in the information above. 

In [4]:
null_dates = df.isnull()

null_dates = null_dates[null_dates.any(axis=1)]
null_dates = null_dates.index.tolist()
print(null_dates)
print(f"Number of null dates: {len(null_dates)}")
print(f"Percentage of null dates: {len(null_dates)/len(df)*100:.2f}%")

['2022-01-17', '2022-02-21', '2022-04-18', '2022-05-30', '2022-06-20', '2022-07-04', '2022-09-05', '2022-11-24', '2023-01-02', '2023-01-16', '2023-02-20', '2023-04-10', '2023-05-01', '2023-05-29', '2023-06-19', '2023-07-04', '2023-09-04', '2023-11-23', '2023-12-26', '2024-01-15', '2024-02-19', '2024-04-01', '2024-05-01', '2024-05-27', '2024-06-19', '2024-07-04', '2024-09-02', '2024-11-28', '2024-12-24', '2024-12-26', '2024-12-31']
Number of null dates: 31
Percentage of null dates: 4.00%


Being the percentage of null dates as low as $4\%$, these registries can be deleted without compromising the integrity of this study. 

In [5]:
df.dropna(inplace=True)

df.to_csv('../data/processed/historical_stock_cleaned.csv')